### PR TITLE
Step03 controller test

### PR DESCRIPTION
**Pull request by mmschaefer**
Finish up Step03ControllerTest

### CODE REVIEW

File: src-test/org/etools/j1939_84/controllers/part1/Step03ControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+        verify(diagnosticReadinessModule).setJ1939(j1939);
+        verify(engineSpeedModule).setJ1939(j1939);
+        verify(vehicleInformationModule).setJ1939(j1939);
+        verify(dataRepository).getObdModules();
+        verify(diagnosticReadinessModule).requestDM5Packets(any(), eq(true));
+        verify(dataRepository).putObdModule(0, obdInfo1);
+        verify(mockListener).addOutcome(1, 3, FAIL, "6.1.3.2.b - The request for DM5 was NACK'ed");
+        verify(reportFileModule).addOutcome(1, 3, FAIL, "6.1.3.2.b - The request for DM5 was NACK'ed");
+        verify(reportFileModule).onResult("FAIL: " + "6.1.3.2.b - The request for DM5 was NACK'ed");
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/controllers/part1/Step03ControllerTest.java]: nit: "+" is not necessary here.

File: src/org/etools/j1939_84/bus/j1939/packets/SupportedSPN.java [hunk 2]
```diff
@@ -38,14 +40,31 @@ public static int parseSPN(int[] data) {
...
+
+        if (!(obj instanceof SupportedSPN)) {
+            return false;
+        }
+
+        SupportedSPN that = (SupportedSPN) obj;
+        return Objects.equals(length, that.length)
+                && Objects.equals(spn, that.spn)
+                && Objects.equals(support, that.support);
+
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/bus/j1939/packets/SupportedSPN.java]: nit: extra line

File: src/org/etools/j1939_84/bus/j1939/packets/SupportedSPN.java [hunk 3]
```diff
@@ -38,14 +40,31 @@ public static int parseSPN(int[] data) {
...
      */
-    SupportedSPN(int[] data) {
+    public SupportedSPN(int[] data) {
         support = data[2] & 0x07;
         spn = SupportedSPN.parseSPN(data);
         length = (byte) (data[3] & 0xFF);
     }
 
+    @Override
+    public boolean equals(Object obj) {
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/bus/j1939/packets/SupportedSPN.java]: hashCode needs to be implemented

File: src/org/etools/j1939_84/model/OBDModuleInformation.java [hunk 4]
```diff
@@ -25,6 +26,22 @@ public OBDModuleInformation(int sourceAddress) {
         this.sourceAddress = sourceAddress;
     }
 
+    @Override
+    public boolean equals(Object obj) {
```
Comment (gumbelmj) [inline src/org/etools/j1939_84/model/OBDModuleInformation.java]: This class needs unit tests.

File: src/org/etools/j1939_84/bus/j1939/packets/SupportedSPN.java (hunk 3 above)
Comment (gumbelmj) [inline src/org/etools/j1939_84/bus/j1939/packets/SupportedSPN.java]: Unit tests need updated

File: src-test/org/etools/j1939_84/bus/j1939/packets/SupportedSPNTest.java [hunk 5]
```diff
@@ -27,6 +38,12 @@ public void testError() {
         assertEquals(expected, instance.toString());
     }
 
+    @Test
+    public void testHashCode() {
+        SupportedSPN instance = new SupportedSPN(new int[] { 0x01, 0x02, 0x1D, 8 });
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/bus/j1939/packets/SupportedSPNTest.java]: This test should make two instances and verify that the hashCode of both is the same.  When two objects are equal, their hashCode must be the same (per the object contract).  (It’s possible that two instances are not equal but the hashCodes just happen to be the same, though that’s rare)

File: src-test/org/etools/j1939_84/bus/j1939/packets/SupportedSPNTest.java [hunk 6]
```diff
@@ -15,6 +17,15 @@
  */
 public class SupportedSPNTest {
 
+    @Test
+    public void testEquals() {
+        SupportedSPN instance = new SupportedSPN(new int[] { 0x01, 0x02, 0x1D, 8 });
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/bus/j1939/packets/SupportedSPNTest.java]: This test should verify that two different instances of SupportedSPN are not the equal because the members that make up the equal functions are different.  Ideally, this would test each member individually, but that might be difficult here.

File: src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    @Test
+    public void testGetTestResultSpns() {
+        assertTrue(instance.getTestResultSpns().equals(instance2.getTestResultSpns()));
+    }
+
+    @Test
+    public void testHashCode() {
+        assertTrue("HashCode", instance.hashCode() == instance.hashCode());
+        assertTrue("HashCode", instance2.hashCode() == instance2.hashCode());
+        assertTrue("HashCode", instance2.hashCode() != instance2.hashCode());
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java]: This can’t possibly pass.

File: src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+    @Test
+    public void testGetObdCompliance() {
+        assertEquals("ObdCompliance", (byte) 4, instance.getObdCompliance());
+        assertEquals("ObdCompliance", instance.getObdCompliance(), instance2.getObdCompliance());
+    }
+
+    @Test
+    public void testGetSupportedSpns() {
+        instance.setSupportedSpns(null);
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java]: This line should be moved down to above the assertNotNull line

File: src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+    @Test
+    public void testGetFreezeFrameSpns() {
+        assertTrue(instance.getFreezeFrameSpns().equals(instance2.getFreezeFrameSpns()));
+    }
+
+    @Test
+    public void testGetObdCompliance() {
+        assertEquals("ObdCompliance", (byte) 4, instance.getObdCompliance());
+        assertEquals("ObdCompliance", instance.getObdCompliance(), instance2.getObdCompliance());
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java]: Using assertEquals here isn’t really right.  You’re asserting that when you call getObdCompliance on instance2, the expected value is what’s returned from the getObdCompliance on instance.  If you’re testing that they are the same, it should be assertTrue(instance.getObdCompliance() == instance2.getObdCompliance()). But that’s not really necessary here at all.  That’s handling when you’re testing equals

File: src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Test
+    public void testGetDataStreamSpns() {
+        assertTrue(instance.getDataStreamSpns().equals(instance2.getDataStreamSpns()));
+    }
+
+    @Test
+    public void testGetFreezeFrameSpns() {
+        assertTrue(instance.getFreezeFrameSpns().equals(instance2.getFreezeFrameSpns()));
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java]: This needs to test that only the FreezeFrame Spns are return, not that they are the same.  This applies for the GetDataStreams as well.

File: src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java (hunk 7 above)
Comment (mmschaefer) [reply src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java]: Copy and paste....

File: src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+    @Test
+    public void testGetSupportedSpns() {
+        assertEquals("SupportedSpn", makeListOfSupportedSPNs(null), instance.getSupportedSpns());
+        instance.setSupportedSpns(null);
+        assertNotNull("SupportedSpn", instance.getSupportedSpns());
+    }
+
+    @Test
+    public void testGetTestResultSpns() {
+        assertTrue(instance.getTestResultSpns().equals(new ArrayList<SupportedSPN>()));
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java]: This needs to tested to return a non-empty list.

File: src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    @Test
+    public void testHashCode() {
+        assertTrue("HashCode", instance.hashCode() == instance.hashCode());
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java]: You need to check that instance.hashCode() == instance2.hashCode()

File: src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        assertTrue("HashCode", instance2.hashCode() == instance2.hashCode());
+
+        OBDModuleInformation instance3 = new OBDModuleInformation(0);
+        instance3.setObdCompliance((byte) 4);
+        instance3.setSupportedSpns(makeListOfSupportedSPNs(new int[] { 1, 3, 5, 7, 9 }));
+        assertTrue("HashCode", instance2.hashCode() != instance3.hashCode());
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java]: in raw cases this can actually fail

File: src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+        SupportedSPN supportedSpn = new SupportedSPN(new int[] { 4, 5, 6, 7, 8 });
+        List<SupportedSPN> supportedSpns = new ArrayList<>();
+        supportedSpns.add(supportedSpn);
+        instance2.setSupportedSpns(supportedSpns);
+        assertFalse(instance.equals(instance2));
+    }
+
+    @Test
+    public void testSetObdCompliance() {
+        assertEquals("SetObdCompliance", instance.getObdCompliance(), instance2.getObdCompliance());
```
Comment (gumbelmj) [inline src-test/org/etools/j1939_84/model/OBDModuleInformationTest.java]: not really sure what's being tested here.
